File: sdk/src/main/java/com/libreshockwave/lingo/decompiler/LingoProperties.java
package com.libreshockwave.lingo.decompiler;

import com.libreshockwave.lingo.Opcode;

import java.util.Map;

/**
 * Static property name maps for Lingo decompilation.
 * Ported from ProjectorRays lingo.cpp.
 */
public final class LingoProperties {

    private LingoProperties() {}

    public static final Map<Opcode, String> BINARY_OP_NAMES = Map.ofEntries(
        Map.entry(Opcode.MUL, "*"),
        Map.entry(Opcode.ADD, "+"),
        Map.entry(Opcode.SUB, "-"),
        Map.entry(Opcode.DIV, "/"),
        Map.entry(Opcode.MOD, "mod"),
        Map.entry(Opcode.JOIN_STR, "&"),
        Map.entry(Opcode.JOIN_PAD_STR, "&&"),
        Map.entry(Opcode.LT, "<"),
        Map.entry(Opcode.LT_EQ, "<="),
        Map.entry(Opcode.NT_EQ, "<>"),
        Map.entry(Opcode.EQ, "="),
        Map.entry(Opcode.GT, ">"),
        Map.entry(Opcode.GT_EQ, ">="),
        Map.entry(Opcode.AND, "and"),
        Map.entry(Opcode.OR, "or"),
        Map.entry(Opcode.CONTAINS_STR, "contains"),
        Map.entry(Opcode.CONTAINS_0_STR, "starts")
    );

    public static final Map<Integer, String> CHUNK_TYPE_NAMES = Map.of(
        1, "char",
        2, "word",
        3, "item",
        4, "line"
    );

    public static final Map<Integer, String> PUT_TYPE_NAMES = Map.of(
        1, "into",
        2, "after",
        3, "before"
    );

    public static final Map<Integer, String> MOVIE_PROPERTY_NAMES = Map.ofEntries(
        Map.entry(0x00, "floatPrecision"),
        Map.entry(0x01, "mouseDownScript"),
        Map.entry(0x02, "mouseUpScript"),
        Map.entry(0x03, "keyDownScript"),
        Map.entry(0x04, "keyUpScript"),
        Map.entry(0x05, "timeoutScript"),
        Map.entry(0x06, "short time"),
        Map.entry(0x07, "abbr time"),
        Map.entry(0x08, "long time"),
        Map.entry(0x09, "short date"),
        Map.entry(0x0a, "abbr date"),
        Map.entry(0x0b, "long date")
    );

    public static final Map<Integer, String> WHEN_EVENT_NAMES = Map.of(
        0x01, "mouseDown",
        0x02, "mouseUp",
        0x03, "keyDown",
        0x04, "keyUp",
        0x05, "timeOut"
    );

    public static final Map<Integer, String> MENU_PROPERTY_NAMES = Map.of(
        0x01, "name",
        0x02, "number of menuItems"
    );

    public static final Map<Integer, String> MENU_ITEM_PROPERTY_NAMES = Map.of(
        0x01, "name",
        0x02, "checkMark",
        0x03, "enabled",
        0x04, "script"
    );

    public static final Map<Integer, String> SOUND_PROPERTY_NAMES = Map.of(
        0x01, "volume"
    );

    public static final Map<Integer, String> SPRITE_PROPERTY_NAMES = Map.ofEntries(
        Map.entry(0x01, "type"), Map.entry(0x02, "backColor"), Map.entry(0x03, "bottom"),
        Map.entry(0x04, "castNum"), Map.entry(0x05, "constraint"), Map.entry(0x06, "cursor"),
        Map.entry(0x07, "foreColor"), Map.entry(0x08, "height"), Map.entry(0x09, "immediate"),
        Map.entry(0x0a, "ink"), Map.entry(0x0b, "left"), Map.entry(0x0c, "lineSize"),
        Map.entry(0x0d, "locH"), Map.entry(0x0e, "locV"), Map.entry(0x0f, "movieRate"),
        Map.entry(0x10, "movieTime"), Map.entry(0x11, "pattern"), Map.entry(0x12, "puppet"),
        Map.entry(0x13, "right"), Map.entry(0x14, "startTime"), Map.entry(0x15, "stopTime"),
        Map.entry(0x16, "stretch"), Map.entry(0x17, "top"), Map.entry(0x18, "trails"),
        Map.entry(0x19, "visible"), Map.entry(0x1a, "volume"), Map.entry(0x1b, "width"),
        Map.entry(0x1c, "blend"), Map.entry(0x1d, "scriptNum"), Map.entry(0x1e, "moveableSprite"),
        Map.entry(0x1f, "editableText"), Map.entry(0x20, "scoreColor"), Map.entry(0x21, "loc"),
        Map.entry(0x22, "rect"), Map.entry(0x23, "memberNum"), Map.entry(0x24, "castLibNum"),
        Map.entry(0x25, "member"), Map.entry(0x26, "scriptInstanceList"),
        Map.entry(0x27, "currentTime"), Map.entry(0x28, "mostRecentCuePoint"),
        Map.entry(0x29, "tweened"), Map.entry(0x2a, "name")
    );

    public static final Map<Integer, String> ANIMATION_PROPERTY_NAMES = Map.ofEntries(
        Map.entry(0x01, "beepOn"), Map.entry(0x02, "buttonStyle"), Map.entry(0x03, "centerStage"),
        Map.entry(0x04, "checkBoxAccess"), Map.entry(0x05, "checkboxType"),
        Map.entry(0x06, "colorDepth"), Map.entry(0x07, "colorQD"), Map.entry(0x08, "exitLock"),
        Map.entry(0x09, "fixStageSize"), Map.entry(0x0a, "fullColorPermit"),
        Map.entry(0x0b, "imageDirect"), Map.entry(0x0c, "doubleClick"), Map.entry(0x0d, "key"),
        Map.entry(0x0e, "lastClick"), Map.entry(0x0f, "lastEvent"), Map.entry(0x10, "keyCode"),
        Map.entry(0x11, "lastKey"), Map.entry(0x12, "lastRoll"), Map.entry(0x13, "timeoutLapsed"),
        Map.entry(0x14, "multiSound"), Map.entry(0x15, "pauseState"),
        Map.entry(0x16, "quickTimePresent"), Map.entry(0x17, "selEnd"),
        Map.entry(0x18, "selStart"), Map.entry(0x19, "soundEnabled"),
        Map.entry(0x1a, "soundLevel"), Map.entry(0x1b, "stageColor"),
        Map.entry(0x1d, "switchColorDepth"), Map.entry(0x1e, "timeoutKeyDown"),
        Map.entry(0x1f, "timeoutLength"), Map.entry(0x20, "timeoutMouse"),
        Map.entry(0x21, "timeoutPlay"), Map.entry(0x22, "timer"),
        Map.entry(0x23, "preLoadRAM"), Map.entry(0x24, "videoForWindowsPresent"),
        Map.entry(0x25, "netPresent"), Map.entry(0x26, "safePlayer"),
        Map.entry(0x27, "soundKeepDevice"), Map.entry(0x28, "soundMixMedia")
    );

    public static final Map<Integer, String> ANIMATION2_PROPERTY_NAMES = Map.of(
        0x01, "perFrameHook",
        0x02, "number of castMembers",
        0x03, "number of menus",
        0x04, "number of castLibs",
        0x05, "number of xtras"
    );

    public static final Map<Integer, String> MEMBER_PROPERTY_NAMES = Map.ofEntries(
        Map.entry(0x01, "name"), Map.entry(0x02, "text"), Map.entry(0x03, "textStyle"),
        Map.entry(0x04, "textFont"), Map.entry(0x05, "textHeight"), Map.entry(0x06, "textAlign"),
        Map.entry(0x07, "textSize"), Map.entry(0x08, "picture"), Map.entry(0x09, "hilite"),
        Map.entry(0x0a, "number"), Map.entry(0x0b, "size"), Map.entry(0x0c, "loop"),
        Map.entry(0x0d, "duration"), Map.entry(0x0e, "controller"),
        Map.entry(0x0f, "directToStage"), Map.entry(0x10, "sound"),
        Map.entry(0x11, "foreColor"), Map.entry(0x12, "backColor"), Map.entry(0x13, "type")
    );

    public static String getName(Map<Integer, String> map, int id) {
        return map.getOrDefault(id, "ERROR");
    }
}

File: cpp/src/lingo/decompiler/LingoProperties.cpp
#include "libreshockwave/lingo/decompiler/LingoProperties.hpp"

namespace libreshockwave::lingo::decompiler {

std::string_view binaryOpName(Opcode opcode) {
    switch (opcode) {
        case Opcode::MUL: return "*";
        case Opcode::ADD: return "+";
        case Opcode::SUB: return "-";
        case Opcode::DIV: return "/";
        case Opcode::MOD: return "mod";
        case Opcode::JOIN_STR: return "&";
        case Opcode::JOIN_PAD_STR: return "&&";
        case Opcode::LT: return "<";
        case Opcode::LT_EQ: return "<=";
        case Opcode::NT_EQ: return "<>";
        case Opcode::EQ: return "=";
        case Opcode::GT: return ">";
        case Opcode::GT_EQ: return ">=";
        case Opcode::AND: return "and";
        case Opcode::OR: return "or";
        case Opcode::CONTAINS_STR: return "contains";
        case Opcode::CONTAINS_0_STR: return "starts";
        default: return "?";
    }
}

std::string_view chunkTypeName(int id) {
    switch (id) {
        case 1: return "char";
        case 2: return "word";
        case 3: return "item";
        case 4: return "line";
        default: return "chunk";
    }
}

std::string_view putTypeName(int id) {
    switch (id) {
        case 1: return "into";
        case 2: return "after";
        case 3: return "before";
        default: return "into";
    }
}

std::string_view moviePropertyName(int id) {
    switch (id) {
        case 0x00: return "floatPrecision";
        case 0x01: return "mouseDownScript";
        case 0x02: return "mouseUpScript";
        case 0x03: return "keyDownScript";
        case 0x04: return "keyUpScript";
        case 0x05: return "timeoutScript";
        case 0x06: return "short time";
        case 0x07: return "abbr time";
        case 0x08: return "long time";
        case 0x09: return "short date";
        case 0x0a: return "abbr date";
        case 0x0b: return "long date";
        default: return "ERROR";
    }
}

std::string_view whenEventName(int id) {
    switch (id) {
        case 0x01: return "mouseDown";
        case 0x02: return "mouseUp";
        case 0x03: return "keyDown";
        case 0x04: return "keyUp";
        case 0x05: return "timeOut";
        default: return "ERROR";
    }
}

std::string_view menuPropertyName(int id) {
    switch (id) {
        case 0x01: return "name";
        case 0x02: return "number of menuItems";
        default: return "ERROR";
    }
}

std::string_view menuItemPropertyName(int id) {
    switch (id) {
        case 0x01: return "name";
        case 0x02: return "checkMark";
        case 0x03: return "enabled";
        case 0x04: return "script";
        default: return "ERROR";
    }
}

std::string_view soundPropertyName(int id) {
    switch (id) {
        case 0x01: return "volume";
        default: return "ERROR";
    }
}

std::string_view spritePropertyName(int id) {
    switch (id) {
        case 0x01: return "type";
        case 0x02: return "backColor";
        case 0x03: return "bottom";
        case 0x04: return "castNum";
        case 0x05: return "constraint";
        case 0x06: return "cursor";
        case 0x07: return "foreColor";
        case 0x08: return "height";
        case 0x09: return "immediate";
        case 0x0a: return "ink";
        case 0x0b: return "left";
        case 0x0c: return "lineSize";
        case 0x0d: return "locH";
        case 0x0e: return "locV";
        case 0x0f: return "movieRate";
        case 0x10: return "movieTime";
        case 0x11: return "pattern";
        case 0x12: return "puppet";
        case 0x13: return "right";
        case 0x14: return "startTime";
        case 0x15: return "stopTime";
        case 0x16: return "stretch";
        case 0x17: return "top";
        case 0x18: return "trails";
        case 0x19: return "visible";
        case 0x1a: return "volume";
        case 0x1b: return "width";
        case 0x1c: return "blend";
        case 0x1d: return "scriptNum";
        case 0x1e: return "moveableSprite";
        case 0x1f: return "editableText";
        case 0x20: return "scoreColor";
        case 0x21: return "loc";
        case 0x22: return "rect";
        case 0x23: return "memberNum";
        case 0x24: return "castLibNum";
        case 0x25: return "member";
        case 0x26: return "scriptInstanceList";
        case 0x27: return "currentTime";
        case 0x28: return "mostRecentCuePoint";
        case 0x29: return "tweened";
        case 0x2a: return "name";
        default: return "ERROR";
    }
}

std::string_view animationPropertyName(int id) {
    switch (id) {
        case 0x01: return "beepOn";
        case 0x02: return "buttonStyle";
        case 0x03: return "centerStage";
        case 0x04: return "checkBoxAccess";
        case 0x05: return "checkboxType";
        case 0x06: return "colorDepth";
        case 0x07: return "colorQD";
        case 0x08: return "exitLock";
        case 0x09: return "fixStageSize";
        case 0x0a: return "fullColorPermit";
        case 0x0b: return "imageDirect";
        case 0x0c: return "doubleClick";
        case 0x0d: return "key";
        case 0x0e: return "lastClick";
        case 0x0f: return "lastEvent";
        case 0x10: return "keyCode";
        case 0x11: return "lastKey";
        case 0x12: return "lastRoll";
        case 0x13: return "timeoutLapsed";
        case 0x14: return "multiSound";
        case 0x15: return "pauseState";
        case 0x16: return "quickTimePresent";
        case 0x17: return "selEnd";
        case 0x18: return "selStart";
        case 0x19: return "soundEnabled";
        case 0x1a: return "soundLevel";
        case 0x1b: return "stageColor";
        case 0x1d: return "switchColorDepth";
        case 0x1e: return "timeoutKeyDown";
        case 0x1f: return "timeoutLength";
        case 0x20: return "timeoutMouse";
        case 0x21: return "timeoutPlay";
        case 0x22: return "timer";
        case 0x23: return "preLoadRAM";
        case 0x24: return "videoForWindowsPresent";
        case 0x25: return "netPresent";
        case 0x26: return "safePlayer";
        case 0x27: return "soundKeepDevice";
        case 0x28: return "soundMixMedia";
        default: return "ERROR";
    }
}

std::string_view animation2PropertyName(int id) {
    switch (id) {
        case 0x01: return "perFrameHook";
        case 0x02: return "number of castMembers";
        case 0x03: return "number of menus";
        case 0x04: return "number of castLibs";
        case 0x05: return "number of xtras";
        default: return "ERROR";
    }
}

std::string_view memberPropertyName(int id) {
    switch (id) {
        case 0x01: return "name";
        case 0x02: return "text";
        case 0x03: return "textStyle";
        case 0x04: return "textFont";
        case 0x05: return "textHeight";
        case 0x06: return "textAlign";
        case 0x07: return "textSize";
        case 0x08: return "picture";
        case 0x09: return "hilite";
        case 0x0a: return "number";
        case 0x0b: return "size";
        case 0x0c: return "loop";
        case 0x0d: return "duration";
        case 0x0e: return "controller";
        case 0x0f: return "directToStage";
        case 0x10: return "sound";
        case 0x11: return "foreColor";
        case 0x12: return "backColor";
        case 0x13: return "type";
        default: return "ERROR";
    }
}

} // namespace libreshockwave::lingo::decompiler
